File: networking/src/main/java/com/jbrunton/networking/services/DescriptiveErrorFactory.java
package com.jbrunton.networking.services;

import com.jbrunton.networking.DescriptiveError;

import java.io.IOException;
import java.lang.annotation.Annotation;
import java.lang.reflect.Type;

import io.reactivex.Observable;
import io.reactivex.functions.Function;
import retrofit2.Call;
import retrofit2.CallAdapter;
import retrofit2.Retrofit;

public class DescriptiveErrorFactory extends CallAdapter.Factory {
    public static CallAdapter.Factory create() {
        return new DescriptiveErrorFactory();
    }

    @Override public CallAdapter<?, ?> get(Type returnType, Annotation[] annotations,
                                           Retrofit retrofit) {
        if (getRawType(returnType) != Observable.class) {
            return null;
        }

        CallAdapter<Object, Observable<?>> delegate =
                (CallAdapter<Object, Observable<?>>) retrofit.nextCallAdapter(this,
                        returnType, annotations);

        return new DescriptiveErrorRxCallAdapter(delegate);
    }

    private static class DescriptiveErrorRxCallAdapter implements CallAdapter<Object, Observable<?>> {
        private final CallAdapter<Object, Observable<?>> delegate;

        private DescriptiveErrorRxCallAdapter(CallAdapter<Object, Observable<?>> delegate) {
            this.delegate = delegate;
        }

        @Override public Type responseType() {
            return delegate.responseType();
        }

        @Override public Observable<?> adapt(Call<Object> call) {
            Observable observable = delegate.adapt(call);
            return observable.onErrorResumeNext(new Function<Throwable, Observable>() {
                @Override public Observable apply(Throwable throwable) throws Exception {
                    if (throwable instanceof IOException) {
                        return Observable.error(
                                new DescriptiveError(
                                        "There was a problem with your connection.",
                                        throwable,
                                        true));
                    } else {
                        return Observable.error(DescriptiveError.from(throwable));
                    }
                }
            });
        }
    }
}

File: networking/src/main/java/com/jbrunton/networking/services/DescriptiveErrorFactory.kt
//package com.jbrunton.networking.services
//
//import com.jbrunton.networking.DescriptiveError
//import io.reactivex.Observable
//import retrofit2.Call
//import retrofit2.CallAdapter
//import retrofit2.Retrofit
//import java.io.IOException
//import java.lang.reflect.Type
//
//class DescriptiveErrorFactory : CallAdapter.Factory() {
//
//    override fun get(returnType: Type, annotations: Array<Annotation>,
//                     retrofit: Retrofit): CallAdapter<*, *>? {
//        if (CallAdapter.Factory.getRawType(returnType) != Observable::class.java) {
//            return null
//        }
//
//        val delegate = retrofit.nextCallAdapter(this,
//                returnType, annotations) as CallAdapter<Any, Observable<Any>>
//
//        return DescriptiveErrorCallAdapter<Any>(delegate)
//    }
//
//    private class DescriptiveErrorCallAdapter<T> constructor(val delegate: CallAdapter<Any, Observable<T>>) : CallAdapter<Any, Observable<T>> {
//
//        override fun responseType(): Type {
//            return delegate.responseType()
//        }
//
//        override fun adapt(call: Call<Any>): Observable<T> {
//            val observable = delegate.adapt(call)
//            return observable.onErrorResumeNext { error: Throwable ->
//                if (error is IOException) {
//                    Observable.error<T>(
//                            DescriptiveError(
//                                    "There was a problem with your connection.",
//                                    error,
//                                    true))
//                } else {
//                    Observable.error<T>(DescriptiveError.from(error))
//                }
//            }
//        }
//    }
//
//    companion object {
//        fun create(): CallAdapter.Factory {
//            return DescriptiveErrorFactory()
//        }
//    }
//}